src/render/src/macros.rs

```diff
@@ -0,0 +1,214 @@
+                    }
+                }
+            )*
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! gfx_graphics_pipeline {
+    ($name:ident {
+        $( $cmp_name:ident: $cmp:ty, )*
+    }) => {
+        #[allow(missing_docs)]
+        pub mod $name {
+            #[allow(unused_imports)]
+            use super::*;
+            use $crate::{pso, handle, Backend, Device, Encoder, Primitive};
+            use $crate::core::pass::{self as cpass, SubpassRef};
+            use $crate::core::{pso as cpso, image as cimg};
+    
+            pub struct Meta<B: Backend> {
+                layout: handle::raw::PipelineLayout<B>,
+                render_pass: handle::raw::RenderPass<B>,
+                pipeline: handle::raw::GraphicsPipeline<B>,
+            }
+
+            pub struct Init<'a, B: Backend> {
+                $( pub $cmp_name: <$cmp as pso::Component<'a, B>>::Init, )*
+            }
+
+            pub struct Data<'a, B: Backend> {
+                $( pub $cmp_name: <$cmp as pso::Component<'a, B>>::Data, )*
+            }
+
+            impl<'a, B: Backend> pso::GraphicsPipelineInit<B> for Init<'a, B> {
+                type Pipeline = Meta<B>;
+
+                fn create(
+                    self,
+                    device: &mut Device<B>,
+                    shader_entries: $crate::core::pso::GraphicsShaderSet<B>,
+                    primitive: Primitive,
+                    rasterizer: pso::Rasterizer
+                ) -> Result<Self::Pipeline, pso::CreationError> {
+                    let mut desc_layouts = Vec::new();
+                    $( desc_layouts.extend(<$cmp as pso::Component<'a, B>>::descriptor_layout(&self.$cmp_name)); )*
+                    let layout = device.create_pipeline_layout_raw(&desc_layouts[..]);
+                    let render_pass = {
+                        let mut attachments = Vec::new();
+                        let mut color_attachments = Vec::new();
+                        $(
+                            if let Some(attach) = <$cmp as pso::Component<'a, B>>::attachment(&self.$cmp_name) {
+                                let attach_id = attachments.len();
+                                attachments.push(cpass::Attachment {
+                                    format: attach.format,
+                                    ops: attach.ops,
+                                    stencil_ops: attach.stencil_ops,
+                                    layouts: attach.required_layout..attach.required_layout,
+                                });
+                                color_attachments.push((attach_id, attach.required_layout));
+                            }
+                        )*
+                        let subpass = cpass::SubpassDesc {
+                            color_attachments: &color_attachments[..],
+                        };
+
+                        // TODO:
+                        let dependency = cpass::SubpassDependency {
```

Comment:
Vulkan's spec (and probably we also) implicitly add subpass dependencies from EXTERNAl -> firstSubpass and lastSubpass -> EXTERNAL, which should cover your use-case also. So this dependencies shouldn't be needed.

Reply:
nice!